### PR TITLE
8264160: Regex \b is not consistent with \w without UNICODE_CHARACTER_CLASS

### PR DESCRIPTION
**Pull request by igraves**
Proposed change in behavior to correct inconsistencies between `\w` and `\b` metacharacters

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed
- [x] Change requires a CSR request to be approved

### Issues
 * [JDK-8264160](https://bugs.openjdk.java.net/browse/JDK-8264160): Regex \b is not consistent with \w without UNICODE_CHARACTER_CLASS
 * [JDK-8282129](https://bugs.openjdk.java.net/browse/JDK-8282129): Regex \b is not consistent with \w without UNICODE_CHARACTER_CLASS (**CSR**)

### Reviewers
 * [Lance Andersen](https://openjdk.java.net/census#lancea) (@LanceAndersen - **Reviewer**)
 * [Brian Burkhalter](https://openjdk.java.net/census#bpb) (@bplb - **Reviewer**) ⚠️ Review applies to 57fef39c5f2a6cc9a9be3ef08b55bcf7167ef2ee
 * [Naoto Sato](https://openjdk.java.net/census#naoto) (@naotoj - **Reviewer**) ⚠️ Review applies to 57fef39c5f2a6cc9a9be3ef08b55bcf7167ef2ee

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/7539/head:pull/7539` \
`$ git checkout pull/7539`

Update a local copy of the PR: \
`$ git checkout pull/7539` \
`$ git pull https://git.openjdk.java.net/jdk pull/7539/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 7539`

View PR using the GUI difftool: \
`$ git pr show -t 7539`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/7539.diff">https://git.openjdk.java.net/jdk/pull/7539.diff</a>

</details>

### CODE REVIEW

File: test/jdk/java/util/regex/RegExTest.java
Comment (LanceAndersen) [inline test/jdk/java/util/regex/RegExTest.java:4557]: There should be an @Test I believe as unless I missed it, this method is not called anywhere else

**Review by LanceAndersen — COMMENTED** [review]
Have not gone through the javadoc changes in detail yet but plan to, only scanned it just now

On a quick skim of the test, it could use some TLC which you might consider doing now or later.  

It would be really good for future maintainers if there was a comment which introduced each test as to what it was trying to exercise (not just a bug number).  Same with some of the private helper methods.

You probably want to include your bug number in the @bug declaration which starts on line 27

**Review by LanceAndersen — COMMENTED** [review]
The changes to Pattern seem OK pending the final CSR approval.   

Looks like your CSR is not quite ready for review as their is no info in the compatibility section.

File: test/jdk/java/util/regex/RegExTest.java
Comment (igraves) [reply test/jdk/java/util/regex/RegExTest.java]: Ah yes, thanks. I had originally written this test with a DataProvider, it proved to be a bit clunky and slower than it is now.

File: test/jdk/java/util/regex/RegExTest.java
Comment (igraves) [reply test/jdk/java/util/regex/RegExTest.java:4563]: Fixed as far back as I can remember for now!

**Review by LanceAndersen — APPROVED** [review]
Thank you for the cleanup.

File: src/java.base/share/classes/java/util/regex/Pattern.java
Comment (bplb) [inline src/java.base/share/classes/java/util/regex/Pattern.java:161]: Looks like there is a superfluous `if` here.

Comment (igraves) [reply src/java.base/share/classes/java/util/regex/Pattern.java]: Thanks!

**Review by bplb — APPROVED** [review]

**Review by naotoj — APPROVED** [review]
LGTM.

**Comment by igraves** [comment]
Something very odd is going on with this PR. I am looking into it.

**Comment by igraves** [comment]
Apologies for the force push. I had a very odd rebase happen that I did not intend. This push rolled it all back out to it's prior state plus a minor change clarifying a regex in the spec.

**Review by LanceAndersen — APPROVED** [review]

**Pull request closed without merging**
